tensorboard/components/tensor_widget/selection.ts

```diff
@@ -0,0 +1,347 @@
+      return null;
+    }
+
+    if (
+      this.slicingSpec.verticalRange === null ||
+      this.slicingSpec.verticalRange[1] === null
+    ) {
+      throw new Error(`Failed to move due to undetermined vertical range.`);
+    }
+
+    if (direction === SelectionMoveDirection.UP) {
+      if (this.rowStart > 0) {
+        this.rowStart--;
+        if (
+          this.slicingSpec.verticalRange != null &&
+          this.rowStart < this.slicingSpec.verticalRange[0]
+        ) {
+          newSlicingSpec = JSON.parse(
+            JSON.stringify(this.slicingSpec)
+          ) as TensorViewSlicingSpec;
+          (newSlicingSpec.verticalRange as [number, number])[0]--;
+          (newSlicingSpec.verticalRange as [number, number])[1]--;
+        }
+      }
+    } else if (direction === SelectionMoveDirection.DOWN) {
+      if (
+        this.slicingSpec.viewingDims != null &&
+        this.slicingSpec.viewingDims[0] != null &&
+        this.rowStart < this.shape[this.slicingSpec.viewingDims[0]] - 1
+      ) {
+        this.rowStart++;
+        if (
+          this.slicingSpec.verticalRange != null &&
+          this.rowStart >= this.slicingSpec.verticalRange[1]
+        ) {
+          newSlicingSpec = JSON.parse(
+            JSON.stringify(this.slicingSpec)
+          ) as TensorViewSlicingSpec;
+          (newSlicingSpec.verticalRange as [number, number])[0]++;
+          (newSlicingSpec.verticalRange as [number, number])[1]++;
+        }
+      }
+    } else if (direction === SelectionMoveDirection.LEFT) {
+      if (this.colStart > 0) {
+        this.colStart--;
+        if (
+          this.slicingSpec.horizontalRange != null &&
+          this.colStart < this.slicingSpec.horizontalRange[0]
+        ) {
+          newSlicingSpec = JSON.parse(
+            JSON.stringify(this.slicingSpec)
+          ) as TensorViewSlicingSpec;
+          (newSlicingSpec.horizontalRange as [number, number])[0]--;
+          (newSlicingSpec.horizontalRange as [number, number])[1]--;
+        }
+      }
+    } else if (direction === SelectionMoveDirection.RIGHT) {
+      if (
+        this.slicingSpec.viewingDims != null &&
+        this.slicingSpec.viewingDims[1] != null &&
+        this.colStart < this.shape[this.slicingSpec.viewingDims[1]] - 1
+      ) {
+        this.colStart++;
+        if (
+          this.slicingSpec.horizontalRange != null &&
+          this.colStart >= (this.slicingSpec.horizontalRange[1] as number)
+        ) {
+          newSlicingSpec = JSON.parse(
+            JSON.stringify(this.slicingSpec)
+          ) as TensorViewSlicingSpec;
+          (newSlicingSpec.horizontalRange as [number, number])[0]++;
+          (newSlicingSpec.horizontalRange as [number, number])[1]++;
+        }
+      }
+    }
+    // Moving the selection causes the selection size to collapse to 1x1.
+    this.rowCount = 1;
+    this.colCount = 1;
+
+    if (newSlicingSpec !== null) {
+      this.slicingSpec = newSlicingSpec;
```

Comment:
I'm still concerned about this: TensorWidgetImpl holds one slicingSpec, and TensorElementSelection holds another.  These stay in sync only if TensorWidgetImpl is the only caller of move(), and if it always updates itself with the return value.  `TensorWidgetImpl.selection` is protected, so that helps; making it private would alleviate this a bit more.  But I think it would be more robust to pass an `updateSlicingSpec()` callback to the constructor, and then call it here.

Furthermore TensorElementSelection should not retain the slicingSpec reference from the constructor at all-- if the parent TensorWidgetImpl updated its version for any reason other than move(), the one here would still point to the old one.  Maybe you always throw out the selection in this case?  Anyway it feels entangled/brittle.